libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt

```diff
@@ -0,0 +1,296 @@
+import kotlin.reflect.KVisibility
+import kotlin.reflect.jvm.isAccessible
+
+/**
+ * Abstract test class that each implementation that follows a builder pattern should implement in a unit test.
+ *
+ * You can read more about the design goals in https://github.com/mapbox/mapbox-navigation-android/issues/2709.
+ *
+ * A target pattern looks more-or-less like this:
+ * ``` kotlin
+ * /**
+ *  * Description of the RequiredExample. This is a data class because it is responsible for
+ *  * transporting data values.
+ *  *
+ *  * @param required description of what to expect from required
+ *  * @param foo description of what to expect from foo
+ *  * @param bar description of what to expect from bar
+ *  */
+ * class Example private constructor(
+ *     val required: String, // required
+ *     val foo: String, // optional with default value
+ *     val bar: Int? // optional without default value
+ * ) {
+ *     /**
+ *      * @return the builder that created the [Example]
+ *      */
+ *     fun toBuilder() = Builder(required).also {
+ *         it.foo(this.foo)
+ *         it.bar(this.bar)
+ *     }
+ *
+ *     override fun equals(other: Any?): Boolean {
+ *         if (this === other) return true
+ *         if (javaClass != other?.javaClass) return false
+ *
+ *         other as Example
+ *
+ *         if (required != other.required) return false
+ *         if (foo != other.foo) return false
+ *         if (bar != other.bar) return false
+ *
+ *         return true
+ *     }
+ *
+ *     override fun hashCode(): Int {
+ *         var result = required.hashCode()
+ *         result = 31 * result + foo.hashCode()
+ *         result = 31 * result + (bar ?: 0)
+ *         return result
+ *     }
+ *
+ *     override fun toString(): String {
+ *         return "Example(required='$required', foo='$foo', bar=$bar)"
+ *     }
+ *
+ *     /**
+ *      * Description for when to use this builder
+ *      *
+ *      * @param required does not have a default value
+ *      */
+ *     class Builder(
+ *         private val required: String
+ *     ) {
+ *         private var foo: String = "default foo value"
+ *         private var bar: Int? = null
+ *
+ *         /**
+ *          * @param foo description of what to expect when writing foo
+ *          * @return Builder
+ *          */
+ *         fun foo(foo: String) =
+ *             apply { this.foo = foo }
+ *
+ *         /**
+ *          * @param bar description of what to expect when writing bar
+ *          * @return Builder
+ *          */
+ *         fun bar(bar: Int?) =
+ *             apply { this.bar = bar }
+ *
+ *         /**
+ *          * Build new instance of [Example]
+ *          * @return Example
+ *          */
+ *         fun build(): Example {
+ *             return Example(
+ *                 required,
+ *                 foo,
+ *                 bar
+ *             )
+ *         }
+ *     }
+ * }
+ * ```
+ */
+abstract class BuilderTest<Implementation : Any, Builder> {
```

Comment:
Did you try making a utility class where you pass it the parameters? When we use inheritance we can only inherit one piece of functionality

Reply:
I did, but in the end, I wanted to make creating the test as easy as possible. If it was a utility class, we would still need to type out all of the test methods manually that then call utilities. Now we only need to type one to kickstart the test, the rest is run automatically.

Reply:
yeah that's fair, still happy with the test. also can see we are still testing failure scenarios https://github.com/mapbox/mapbox-navigation-android/pull/3382/files#diff-606a7ab2c9307301f681d1a1cfa57e94R57